src/main/java/com/aws/greengrass/status/FleetStatusService.java

```diff
@@ -298,38 +306,79 @@ public void schedulePeriodicFleetStatusDataUpdate(boolean isDuringConnectionResu
-
             // Evaluate overall status based on current state of all components
             Instant now = Instant.now();
             AtomicReference<OverallStatus> overAllStatus = new AtomicReference<>(OverallStatus.HEALTHY);
             this.kernel.orderedDependencies().forEach(service -> {
                 serviceFssTracksMap.put(service, now);
                 overAllStatus.set(getOverallStatusBasedOnServiceState(overAllStatus.get(), service));
             });
-
-            uploadFleetStatusServiceData(erroredComponentSet, overAllStatus.get(), null,
+            // Not reporting status of other components in errored component message because a deployment could
+            // be in-progress and other component's fleet config arn may not have updated
+            Set<GreengrassService> erroredComponent = new HashSet<>();
+            erroredComponent.add(greengrassService);
+            if (isDeploymentInProgress.get()) {
+                uploadFleetStatusServiceData(erroredComponent, overAllStatus.get(), null,
+                        Trigger.COMPONENT_STATUS_CHANGE);
+                return;
+            }
+            // Report status of other components, in case recovery duration takes too long and other components need
+            // to wait until all components to reach terminal state to update status
+            uploadFleetStatusServiceData(updatedGreengrassServiceSet, overAllStatus.get(), null,
                     Trigger.COMPONENT_STATUS_CHANGE);
             return;
         }
-
         synchronized (updatedGreengrassServiceSet) {
             updatedGreengrassServiceSet.add(greengrassService);
         }
 
-        // if there is no ongoing deployment and we encounter a BROKEN component, update the fleet status as UNHEALTHY.
-        if (!isDeploymentInProgress.get() && newState.equals(State.BROKEN)) {
-            synchronized (updatedGreengrassServiceSet) {
-                uploadFleetStatusServiceData(updatedGreengrassServiceSet, OverallStatus.UNHEALTHY,
-                        null, Trigger.COMPONENT_STATUS_CHANGE);
+        if (!isDeploymentInProgress.get()) {
+            // if there is no ongoing deployment and we encounter a BROKEN component,
+            // update the fleet status as UNHEALTHY.
+            if (newState.equals(State.BROKEN)) {
+                synchronized (updatedGreengrassServiceSet) {
+                    uploadFleetStatusServiceData(updatedGreengrassServiceSet, OverallStatus.UNHEALTHY,
+                            null, Trigger.COMPONENT_STATUS_CHANGE);
+                }
+            }
+            // Check and update component status when a component reaches to terminal state
+            if (greengrassService.reachedDesiredState() || newState.equals(State.BROKEN)) {
+                handleServiceReachedTerminalState();
             }
         }
     }
 
+    private void handleServiceReachedTerminalState() {
+        //1. Check all other component status
```

Comment:
this part of checking all services in terminal state can move to Kernel class

Reply:
Updated